### DISCUSSION TITLE
Issues regarding models using moment-to-moment drift rates

### DISCUSSION THREAD

**Discussion by ding9889** [mwshinn/PyDDM#104] [Ask for help] [answered]
Hi Max and the PyDDM community,

Thanks for developing this awesome library!

I'm recently doing some simulation using pyddm. In particular, I'm currently running a model which assumes the drift rates is a function of a moment-to-moment signal presented in the experiment. Everything works fine with the model_gui (btw the GUI is really cool and helpful!), but when I try to solve the model to generate some simulated data, I got an error message:

TypeError: can't multiply sequence by non-int of type 'Fittable'

I'm guessing this is related to the moment-to-moment drifts, seems like rather than mutiplying the signal at a particular moment by the drift scale, the program multiplies the whole array with the drfit scale. But I'm a bit confused since I passed the **conditions** parameters to the **solve()** function in exactly the same way as how I passed it to the **model_gui()** function.

Here are the codes of the model specification:

def find_drift(t, slope_vector):
    t_bin = int(t // 1) # 1000 ms binning
    t_bin = min(len(slope_vector)-1, t_bin) # Never go past the final arousal bin
    return slope_vector[t_bin]

model = pyddm.gddm(drift=lambda t,slope_scale,slope_vector : find_drift(t, slope_vector)*slope_scale,
                   bound="b", starting_position="x0", nondecision="ndt", noise=1,
               parameters={"slope_scale": (0, 2), "b":(.5,10), "x0": (-.8, .8), "ndt": (.1, .5)},
               conditions=["slope_vector"], choice_names=("buy", "sell"), dx=.01, dt=.01, T_dur=10)

Here are the codes for calling model_gui and solving the model, the model_gui works perfectly well, but then I got the TypeError while solving the model:

model_gui(model, conditions={"slope_vector": [(0, 2, 0, 2)]})

sol = model.solve(conditions={"slope_vector": [(0, 2, 0, -2)]})

I was wondering why this is happening and how can I possibly fix it. Any advices and comments would be appreciated!

Best wishes,
Yuyang

**Comment by mwshinn** [comment, marked as answer]
Hi Yuyang,

It looks like there are two issues here, and neither are related to moment-to-moment drifts!  Both are related to how you call model.solve.  In short, you cannot call model.solve the same way you call model_gui.

First, when you call model.solve, you need to specify a single value to solve with for each condition.  Here, the value you want is (0, 2, 0, -2).  Instead, you are specifying a list of values which only contains one element, namely, [(0, 2, 0, -2)].  By contrast, in model_gui, you **can** pass a list of elements because then it will average across them and allow you to select them in the list of conditions.  I am assuming this was unintended and you meant to use (0, 2, 0, -2), but if you actually do want to solve multiple conditions at once (i.e. [(0, 2, 0, -2)] was intended), you can use pyddm.solve_partial_conditions to get the same behaviour as model_gui.

The second issue is that you have not yet fit the parameters of your model before you call model.solve.  So you will get an error because the parameters are still unspecified, since they haven't yet been fit to data.  Once you fit the model to data and remove the surrounding [] on your slope_vector argument, it should work!

**Reply by ding9889** [reply]
Hi Max, thank you so much for the super detailed explanation, now it works!

**Reply by ding9889** [reply]
Hi Max,

Thanks for the answers, that really helps a lot! 

I managed to solve the model after fitting the parameters and removing the brackets, but then I triggered an **AssertionError** with no message when trying to call the solution.sample() function to draw simulated data from the RT distribution. 

The error is triggered while the function trying to construct the **Sample** object. Specifically, it was triggered in the Line 109 of **sample.py**:

assert v[0] in NDArray(d=1)

which, according to the comments around, is meant to check if the sample has the correct shape but I'm not sure why this is happening...I also tried to replicate the example in the cookbook:

df = pandas.DataFrame({"choice_correct": [1, 0, 1], "arousal": [(.5, .7, .8, .9), (.2, .3, .4, .1, .1), (1.1, 1.3)], "RT": [.65, .94, .30]})
samp = pyddm.Sample.from_pandas_dataframe(df, rt_column_name="RT", choice_column_name="choice_correct")

def find_drift(t, arousal_vec):
    t_bin = int(t // .2) # 200 ms binning
    t_bin = min(len(arousal_vec)-1, t_bin) # Never go past the final arousal bin
    return arousal_vec[t_bin]

m = pyddm.gddm(drift=lambda t,arousal_scale,arousal : find_drift(t, arousal)*arousal_scale,
               parameters={"arousal_scale": (0, 2)},
               conditions=["arousal"])

and fit the model using artificial data:

m.fit(samp, verbose=False)

then I managed to solve the model using one set of arousal signals by:

s = m.solve(conditions={"arousal": (.5, .7, .8, .9)})

everything worked well until here, but then I got the same AssertionError when called the **sample()** function like this:

samp = s.sample(100)

I am a bit confused since I am not super clear on what is **v** supposed to hold and why the **v[0]** is not in the demanded shape. Would you mind telling me what are the possible reasons for this error and how can I fix them?

Many thanks,
Yuyang

**Reply by mwshinn** [reply]
Thanks for reporting the issue, this was a bug in PyDDM.  It is fixed with the latest commit, and the fix will be included in the next major update.  Until then, you can either install PyDDM from source from the latest commit or else edit the following lines in the file solution.py in your PyDDM installation: https://github.com/mwshinn/PyDDM/commit/408775bf086a35181ed64836cab00b8f9d000bae

**Reply by ding9889** [reply]
Great! Thank you so much Max!
